File: C++/Solution227.cpp
#include<iostream>
#include<vector>
#include<stack>

using namespace std;

class Solution227 {
public:
    int calculate(string s) {
        int number = 0;
        char presign = '+';
        stack<int> stk;
        for(int i = 0; i < s.length(); i++){
            if(s[i] >= '0' && s[i] <= '9'){
                number = 10 * number + (s[i] - '0');
            }
            if(s[i] == ' ' && i != s.length() - 1){
                continue;
            }
            if(s[i] == '+' || s[i] == '-' || s[i] == '*' || s[i] == '/' || i == s.length() - 1){
                int cur = stk.empty() ? 0 : stk.top();
                switch(presign){
                    case '+':
                    stk.push(number);
                    break;

                    case '-':
                    stk.push(-number);
                    break;

                    case '*':
                    stk.pop();
                    stk.push(cur * number);
                    break;

                    case '/':
                    stk.pop();
                    stk.push(cur / number);
                    break;

                    default:
                    break;
                }
                number = 0;
                presign = s[i];
            }
        }
        int res = 0;
        while(!stk.empty()){
            res += stk.top();
            stk.pop();
        }
        return res;
    }
};

File: Java/stack/Solution227.java

import java.util.Stack;
import java.util.List;
import java.util.LinkedList;

class Solution227 {
    public int calculate(String s) {
        
        // 保存上一个符号，初始为 +
        char sign = '+';
        Stack<Integer> numStack = new Stack<>();
        // 保存当前数字，如：12是两个字符，需要进位累加
        int num = 0;
        int result = 0;
        for(int i = 0; i < s.length(); i++){
            char cur = s.charAt(i);
            if(cur >= '0'){
                // 记录当前数字。先减，防溢出
                num = num*10 - '0' + cur;
            }
            if((cur < '0' && cur !=' ' )|| i == s.length()-1){
                // 判断上一个符号是什么
                switch(sign){
                    // 当前符号前的数字直接压栈
                    case '+': numStack.push(num);break;
                    // 当前符号前的数字取反压栈
                    case '-': numStack.push(-num);break;
                    // 数字栈栈顶数字出栈，与当前符号前的数字相乘，结果值压栈
                    case '*': numStack.push(numStack.pop()*num);break;
                    // 数字栈栈顶数字出栈，除于当前符号前的数字，结果值压栈
                    case '/': numStack.push(numStack.pop()/num);break;
                }
                // 记录当前符号
                sign = cur;
                // 数字清零
                num = 0;
            }
        }
        // 将栈内剩余数字累加，即为结果
        while(!numStack.isEmpty()){
            result += numStack.pop();
        }
        return result;
    }

    public static void main(String[] args){
        String s1 = "1 + 2* 5";
        Solution227 s = new Solution227();
        System.out.println(s.calculate(s1));
    }
}